integrations/kube-agent-updater/pkg/img/cosign.go

```diff
@@ -0,0 +1,119 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package img
+
+import (
+	"context"
+	"crypto"
+	"encoding/hex"
+
+	"github.com/docker/distribution/reference"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/gravitational/trace"
+	"github.com/opencontainers/go-digest"
+	"github.com/sigstore/cosign/v2/pkg/cosign"
+	ociremote "github.com/sigstore/cosign/v2/pkg/oci/remote"
+	"github.com/sigstore/sigstore/pkg/cryptoutils"
+	"github.com/sigstore/sigstore/pkg/signature"
+)
+
+const hashAlgo = crypto.SHA256
+
+type cosignKeyValidator struct {
+	verifier        signature.Verifier
+	skid            []byte
+	name            string
+	registryOptions []ociremote.Option
+}
+
+// Name returns the validator name, it is composed of a pretty name chosen at creation
+// and its public SubjectKeyID hex-encoded.
+func (v *cosignKeyValidator) Name() string {
+	prettySKID := hex.EncodeToString(v.skid)
+	return v.name + "-" + prettySKID
+}
+
+func (v *cosignKeyValidator) ValidateAndResolveDigest(ctx context.Context, image reference.NamedTagged) (NamedTaggedDigested, error) {
+	// No RegistryClientOpts for now, this means no private repo support
+	// This might also be useful for local tetsing (running a test registry)
+	checkOpts := &cosign.CheckOpts{
+		RegistryClientOpts: v.registryOptions,
+		Annotations:        nil,
+		ClaimVerifier:      cosign.SimpleClaimVerifier,
+		SigVerifier:        v.verifier,
+		IgnoreTlog:         true, // TODO: should we keep this?
+	}
+
+	ref, err := NamedTaggedToDigest(image)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	verified, _, err := cosign.VerifyImageSignatures(ctx, ref, checkOpts)
+	if err != nil {
+		return nil, trace.Trust(err, "failed to verify image signature")
+	}
+	if len(verified) == 0 {
+		return nil, &trace.TrustError{Message: "cannot validate image: no valid signature found"}
+	}
+
+	// There are legitimate use-cases where the signing reference is not the same
+	// as the img we're pulling from: img promoted to an internal registry,
+	// cache, registry migration, ...
+	// Thus, we take only take the digest from the signature and use it with our base img.
+	// This is what cosign.SimpleClaimVerifier is doing anyway.
+	digestedImage := NewImageRef(ref.RegistryStr(), ref.RepositoryStr(), image.Tag(), digest.Digest(ref.DigestStr()))
```

Comment:
does `NewImageRef` need to be exported?

Reply:
For now yes, because it is used to create a MaintenanceTriggerMock in `main.go`. When we'll have everything wired together we might be able to make this private.